packages/workspace/src/node/default-workspace-server.ts

```diff
@@ -101,14 +101,16 @@ export class DefaultWorkspaceServer implements WorkspaceServer, BackendApplicati
         return this.root.promise;
     }
 
-    async setMostRecentlyUsedWorkspace(uri: string): Promise<void> {
+    async setMostRecentlyUsedWorkspace(rawUri: string): Promise<void> {
+        const uri = FileUri.fsPath(rawUri);
```

Comment:
Hi, can you explain what this change does? I wonder why you name the variable `uri` when it's a path. Any concerns about backward compatibility?

Reply:
The current code treats `recentworkspaces.json` as a record of unique past workspaces, so it needs to be sure that the strings it's dealing with are consistently formatted. In principle, it would be fine if they were all full stringified URI's - then the file is less human readable, but that isn't really its purpose - if you strongly prefer that solution. The hitch there is that  we can't guarantee that the caller has sent the the scheme, so we'd have to be OK with a default of `file`. Re: the name `uri`, that was mostly to limit the diff (the subsequent code after the normalization doesn't have to be changed), but you're right, `path` would be more accurate.

Reply:
> The hitch there is that we can't guarantee that the caller has sent the the scheme

I am not sure if Theia's default workspace server impl has to be fault-tolerant in this case. The code you have changed worked fine in the last four years or so.

When does the FE send a URI without the scheme? Did you manage to figure it out? Maybe one or two tests could easily show that you're right, and that's the fix. But I cannot agree on the commit message, title, and PR description:

> by normalizing URI's 

You're not normalizing the URI, but changing the URI to an FS path.

If the `recentworkspaces.json` file always stored URIs, it would be great not to switch to FS paths. No changelog, no tests, nothing just three steps I can do to verify the UI.

After checking out your branch and trying it on Windows, this is the `recentworkspaces.json` file content on my disk:

```
{"recentRoots":["\\","\\Users\\kittaakos\\dev\\git\\theia\\doc"]}
```

If I open a folder as a workspace, then the `rawUri` is `file:///c%3A/Users/kittaakos/scoop`, the FS path that you named `uri` is `c:\\Users\\kittaakos\\scoop`. We set this as the root URI, and nothing happens. Theia opens without a workspace, or open on my C: root. Open recent workspaces show this:

![Screenshot 2022-04-13 141026](https://user-images.githubusercontent.com/1405703/163177123-a84b11e9-4872-4423-814d-efb941b566f2.png)

I propose revisiting this PR. I can help with the verification on Windows. What do you think?

Reply:
> I am not sure if Theia's default workspace server impl has to be fault-tolerant in this case. 

The trouble is that the framework failed to implement a good serialization for URI's, and so there are a lot of signatures for exchanges between the backend and frontend like the one I'm changing:

```
                uri: string
                ~~~  ******
What we hope to get  What we can be (fairly) sure we'll get
```

Because we can't guarantee that we'll get correctly formatted full URI's odd things can happen. #11013 is one such, where a mix of stringified URI's and stringified paths lets a duplicate. It looks like the code I've written causes more problems, as per @marol1210's comment, and I'm grateful for your checks on Windows. However, that seems like a problem with the `FileUri.fsPath` implementation: it's spitting out garbage when it should be able to consistently format intelligible paths. We can dodge that problem (though it should be fixed), by using `URI.toString()`, and I'm happy to do that, but I would still say that the backend workspace service should take responsibility for normalizing what it receives to avoid problems like #11013. What do you thonk?

Reply:
> we can't guarantee that we'll get correctly formatted full URI's, odd things can happen

Why not try to parse the string URI to a URI object? If it fails, it fails; the FE has sent garbage.

I do not understand why you want to use FS paths as the workspace root URI. I might not be aware of some recent changes in Theia; please help me. Let's assume the following happy path: this method receives a full valid URI on Windows as a string: `'file:///c%3A/path/to/resource'`, then you convert it to FS path string: `'c:\\path\\to\\resource'`, and we set this string as the workspace URI.

I think I get what you want to achieve; the `file:///path/to/resource` URI on UNIX is the same as the FS path if you drop the scheme: `/path/to/resource`, but it won't work on Windows. Maybe you could try to convert back to a URI and see if it fixes your problem. (I think VS Code can parse URI without a scheme, but then it will be converted to `file`. Please check if the back conversion solves your issue.) Something like this:

```ts
    async setMostRecentlyUsedWorkspace(rawUri: string): Promise<void> {
        const uri = FileUri.create(FileUri.fsPath(rawUri)); // <----
        this.root = new Deferred();
        this.root.resolve(uri);
        // and the rest of the method
    }
```

To sum up, I think we cannot switch from URI to FS path like this. This would be a breaking change for all Windows users. Thoughts? Thanks!

Reply:
> Why not try to parse the string URI to a URI object? If it fails, it fails; the FE has sent garbage.

As I've said, I am happy to use URI's:

> ...it would be fine if they were all full stringified URI's...
> ...by using URI.toString()...

I'll make the changes.